compiler/rustc_middle/src/middle/privacy.rs

```diff
@@ -118,10 +152,56 @@ impl<Id: Hash + Eq + Copy> AccessLevels<Id> {
         self.map.insert(id, effective_vis);
     }
 
+    pub fn calculate_effective_vis(
+        &self,
+        current_id: Id,
+        current_vis: Visibility,
+        module_id: Id,
+        tag: AccessLevel,
+        tree: impl DefIdTree,
+    ) -> Option<EffectiveVisibility> {
+        if let Some(inherited_effective_vis) = self.get_effective_vis(module_id) {
+            let mut current_effective_vis =
+                self.get_effective_vis(current_id).copied().unwrap_or_default();
+            for level in [
+                AccessLevel::Public,
+                AccessLevel::Exported,
+                AccessLevel::Reachable,
+                AccessLevel::ReachableFromImplTrait,
+            ] {
+                if level <= tag {
+                    if let Some(nearest_available_vis) =
+                        inherited_effective_vis.nearest_available(level)
+                    {
+                        let calculated_effective_vis = match current_vis {
+                            Visibility::Public => nearest_available_vis,
+                            Visibility::Restricted(_) => {
+                                if current_vis.is_at_least(nearest_available_vis, tree) {
```

Comment:
Looks like you don't need the `match current_vis`, this logic is correct for both `Visibility::Public` and `Visibility::Restricted`.